### DISCUSSION TITLE
Async/Await in IDataListSource

### DISCUSSION THREAD

**Discussion by Nikkelmann** [leekelleher/umbraco-contentment#259] [Question]
Ran into a problem today where I had an async data source that I wanted to hook up to contentment with the IDataListSource, but the interface does not allow async method calls.

Would it be possible to implement this or is this a restraint by Umbraco (v10)?

**Comment by leekelleher** [comment]
@Nikkelmann Honestly, I _(currently)_ have no idea. I probably need some education on how it could be implemented using async. 😬😅

**Reply by Matthew-Wise** [reply]
If I found some time which branch would be the best to use for a POC and/or PR?

Most code wouldnt need async/await I think mainly it would be adding Task<T> as the returns

**Reply by leekelleher** [reply]
@Matthew-Wise Thank you for the offer, but I doubt it's worth the effort. 

The reason being is that the `IDataListSource.GetItems` method is called from with the `ConfigurationEditor.ToValueEditor` method, which isn't async itself... so even if we wanted to add async support to Data List, it'd be faking it anyway. 😕 
https://github.com/leekelleher/umbraco-contentment/blob/4.4.4/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListConfigurationEditor.cs#L114

The WIP branch for the upcoming Data Picker (PR #297) is different in that the items are populated in a separate fetch request from the client/backoffice, so can be done as async.  To make the Data List editor do the same would require a total rewrite of all the list-editors, (which I'm not prepared to do at this point, with the v14 new backoffice rewrite looming).

**Reply by Matthew-Wise** [reply]
I think its possible with just an interface change to allow people who need to call the datalist in there own code by changing the interface at least for 10+

 IEnumerable<DataListItem> GetItems(Dictionary<string, object> config); => GetItemsAsync(config).GetAwaiter().GetResult();

 Task<IEnumerable<DataListItem>> GetItemsAsync(Dictionary<string, object> config) => Task.FromResult(GetItems(config));


GetAwaiter().GetResults() is never ideal. So would be good to push people to implement both methods. If they want to use Async.

**Reply by leekelleher** [reply]
Sure, that could be done... but the async version wouldn't be truly async, as the caller is Umbraco's [`ConfigurationEditor.ToValueEditor` method](https://github.com/umbraco/Umbraco-CMS/blob/release-12.0.1/src/Umbraco.Core/PropertyEditors/ConfigurationEditor.cs#L137) - which isn't async, _(as far as I'm aware)_.

I wouldn't want to mislead folk in thinking that Data List was async, when it'd be faking it.

**Reply by leekelleher** [reply]
My current thought is that Data List for v14 should be a ground-up rewrite (as opposed to a directly code migration/port), where it could make use of advancements in the platforms, _(given that Data List was originally designed for Umbraco 8/.NET Framework)._
